### PR TITLE
Restrict use of 'size' on components of expandable connectors

### PR DESCRIPTION
**Pull request by henrikt-ma**
Implements the part of the decision in #2560 that wasn't included in #2589.

### CODE REVIEW

**Comment by henrikt-ma** [comment]
> We really need to clarify the "declared" notion inside expandable connectors because the language used here although informally understandable conflicts with parts of 9.1.3

Yes, that sounds like a good idea, but I'm not sure this PR is introducing any unclarity that didn't exist before?

**Comment by HansOlsson** [comment]
I don't recall that we fully decided on this, I thought it was more an additional topic that we should consider.
There could be complicated implications.

**Comment by henrikt-ma** [comment]
> I don't recall that we fully decided on this, I thought it was more an additional topic that we should consider.
> There could be complicated implications.

OK.  I was trying to be GitHub lean by just using a PR instead of opening an issue where I'd make a first suggestion for how to resolve, plus a PR doing the exact same thing as described in the issue.  I expect this PR to be discussed just like any other issue before being merged.

@qlambert-pro, are you in a good position to test implement rejection of these `size` calls?

**Comment by HansOlsson** [comment]
Ok, it wasn't clear to me that this just a proposal. I would propose turning it into a draft to avoid accidental merges.

What I could imagine is that some have a number of models using an expandable connectors and one source that automatically generates values for all of them - using size of the array in the bus.

**Comment by henrikt-ma** [comment]
> Ok, it wasn't clear to me that this just a proposal. I would propose turning it into a draft to avoid accidental merges.

Done.  (… but I'm not sure this should be the mechanism for avoiding accidental merges.  To me, marking it as a draft would mean that it is even too early to start a discussion; I made it not a draft to start the discussion with a concrete proposal for how to resolve the issue.)

**Comment by henrikt-ma** [comment]
> What I could imagine is that some have a number of models using an expandable connectors and one source that automatically generates values for all of them - using size of the array in the bus.

Sure, there will always be potential uses, but I think we should look for more concrete and compelling examples where it is clear that using `size` is the right mechanism for the job.

**Comment by HansOlsson** [comment]
> Done. (… but I'm not sure this should be the mechanism for avoiding accidental merges. To me, marking it as a draft would mean that it is even too early to start a discussion; I made it not a draft to start the discussion with a concrete proposal for how to resolve the issue.)

I agree in general, but in this case I believe the formulation of the first post and the review made it a bit unclear.

**Comment by HansOlsson** [comment]
> Sure, there will always be potential uses, but I think we should look for more concrete and compelling examples where it is clear that using `size` is the right mechanism for the job.

True the example was sort of contrived, and the most clear argument against using size on arrays in connectors is that you are only supposed to "read" from variables in buses based on what you connect to the bus locally and the size-information breaks that.

One related issue is whether the following (based on https://specification.modelica.org/master/connectors-and-connections.html#expandable-connectors ) is legal:

```
expandable connector EngineBus
end EngineBus ;
block Sensor
  RealOutput speed ;
end Sensor ;
model Engine
  parameter Integer n=1;
  EngineBus bus ;
  Sensor sensor ;
   RealOutput sensorSpeeds[:] ;
equation
  connect ( bus.speed[n] , sensor.speed ) ; // provides signal
  connect( bus.speed, sensorSpeeds); // Reads the entire array
public
  Real s[:]= bus.speed; // Reads in another way.
end Engine ;
```
If we forbid size for arrays in buses I think these should also be illegal, and since bus-arrays may have extra elements I don't think these are good style.

**Comment by henrikt-ma** [comment]
> If we forbid size for arrays in buses I think these should also be illegal, and since bus-arrays may have extra elements I don't think these are good style.

Yes, this is a really good point.

File: chapters/connectors.tex [hunk 1]
```diff
@@ -130,26 +128,25 @@ \subsection{Expandable Connectors}\doublelabel{expandable-connectors}
...
-  If the variable on the other side of the connect-equation is input or
-  output the new component will be either input or output to satisfy the
-  restrictions in \autoref{restrictions-of-connections-and-connectors} for a non-expandable connector.
-  \begin{nonnormative}
-  If the existing side refers to an inside connector (i.e.\ a connector of a component) the new variable will copy its causality, i.e.\ input
-  if input and output if output, since the expandable connector must be an outside connector.
-  \end{nonnormative}
-  For an array the input/output property can be deduced separately for each array element.
-\end{itemize}
+  \begin{itemize}
```
Comment (HansOlsson) [inline chapters/connectors.tex:131]: i don't understand why these lines are indented one space; and that also makes it hard to see the diff.

File: chapters/operatorsandexpressions.tex [hunk 2]
```diff
@@ -1317,19 +1317,16 @@ \subsection{Parameter Expressions}\doublelabel{parameter-expressions}
   \lstinline!edge!, \lstinline!change!, \lstinline!sample!, and \lstinline!pre!, a function or operator with parameter
   subexpressions is a parameter expression.
 \item
-  Some function calls are parameter expressions even if the arguments
-  are not:
-
+  Some function calls are parameter expressions even if the arguments are not:
```
Comment (HansOlsson) [inline chapters/operatorsandexpressions.tex:1320]: as far as I understand this is an unrelated formatting change. I couldn't see anything related to expandable connectors here.

File: chapters/connectors.tex [hunk 3]
```diff
@@ -246,16 +242,10 @@ \subsection{Expandable Connectors}\doublelabel{expandable-connectors}
...
 \end{itemize}
-  If a connect equation references a potentially present variable, or
-  variable element, in an expandable connector the variable or variable
-  element is marked as being present, and due to the paragraphs above it
-  is possible to deduce whether the bus variable shall be treated as
-  input, or shall be treated as output in the connect equation. That
-  input or output prefix is added if no input/output prefix is present
-  on the declaration
+
+If a connect equation references a potentially present variable, or variable element, in an expandable connector the variable or variable element is marked as being present, and due to the paragraphs above it is possible to deduce whether the bus variable shall be treated as input, or shall be treated as output in the connect equation.  That \lstinline!input! or \lstinline!output! prefix is added if no \lstinline!input!/\lstinline!output! prefix is present on the declaration.
```
Comment (HansOlsson) [inline chapters/connectors.tex:248]: as far as I understand this is an unrelated formatting change, or is there some other change I'm not seeing.

File: chapters/connectors.tex (hunk 1 above)
Comment (henrikt-ma) [reply chapters/connectors.tex:131]: The problem is that the source gets very confusing to real when nested item lists are using the same indentation.

File: chapters/operatorsandexpressions.tex (hunk 2 above)
Comment (henrikt-ma) [reply chapters/operatorsandexpressions.tex:1320]: I'm afraid the commit message is all I have to say:
> Tidy up function calls that are parameter expressions
>
> Stumbled upon these when looking for specification of 'size'.

File: chapters/connectors.tex (hunk 3 above)
Comment (henrikt-ma) [reply chapters/connectors.tex:248]: Yes, only cleanup related to expandable connectors in general (this is the `\lstinline` part):
> Improve readability of source of nested item list on expandable connectors
>
> Also throwing in some missing \lstinline.

I agree that `\lstinline` cleanup doesn't really belong here.  It's just that I'm not sure we'll ever actually sit down and fix such formatting for the entire specification at once.  Therefore, I was trying the approach of gradually fixing the situation by improving things where work is done at the moment.  Perhaps we need another strategy…

Comment (henrikt-ma) [reply chapters/connectors.tex:248]: The stuff that doesn't belong to this PR has been extracted to #2606.  Getting that on _master_ and then resolving the conflicts should free this PR of unrelated cleanup.

Comment (HansOlsson) [reply chapters/connectors.tex:248]: Ok, makes sense.

File: chapters/operatorsandexpressions.tex (hunk 2 above)
Comment (henrikt-ma) [reply chapters/operatorsandexpressions.tex]: Fixed via #2606.  Resolving.

**Comment by henrikt-ma** [comment]
After the recent merge of _master_, the sum of all changes introduced by this PR is now more to the point.

I am thinking about when would be the right time to remove the draft state of a PR like this, that isn't obviously just implementing something decided at a meeting.  I suppose we need to set up our own interpretation of what being in draft state means…  For now, I'll leave it in draft state as I think we really shouldn't merge it until it has been more actively exposed to a larger audience (the list of issue participants is only 3 long at this point).  Going into summer vacation times, I don't think we should push this forward by setting up a vote — we might as well take a break and wait until the next web meeting.

**Comment by HansOlsson** [comment]
This wasn't discussed at the phone meeting, or at least no-one found it problematic.
I'll give it until next week as well.

**Comment by henrikt-ma** [comment]
I just noted that **continuous-integration/jenkins/pr-merge** didn't protect us against a broken pdfLaTeX (due to use of `\autoref`)?

@HansOlsson: It seems like the thing one needs to do when dealing with such an old branch is to push an empty commit in order to re-trigger the checks before hitting _Merge_.
